extensions/smallrye-fault-tolerance/deployment/src/main/java/io/quarkus/smallrye/faulttolerance/deployment/FaultToleranceScanner.java

```diff
@@ -0,0 +1,167 @@
+package io.quarkus.smallrye.faulttolerance.deployment;
+
+import java.lang.annotation.Annotation;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.Consumer;
+
+import org.eclipse.microprofile.faulttolerance.Asynchronous;
+import org.eclipse.microprofile.faulttolerance.Bulkhead;
+import org.eclipse.microprofile.faulttolerance.CircuitBreaker;
+import org.eclipse.microprofile.faulttolerance.Fallback;
+import org.eclipse.microprofile.faulttolerance.Retry;
+import org.eclipse.microprofile.faulttolerance.Timeout;
+import org.jboss.jandex.AnnotationInstance;
+import org.jboss.jandex.ClassInfo;
+import org.jboss.jandex.DotName;
+import org.jboss.jandex.IndexView;
+import org.jboss.jandex.MethodInfo;
+
+import io.quarkus.arc.processor.AnnotationStore;
+import io.quarkus.deployment.builditem.AnnotationProxyBuildItem;
+import io.quarkus.gizmo.ClassOutput;
+import io.smallrye.common.annotation.Blocking;
+import io.smallrye.common.annotation.NonBlocking;
+import io.smallrye.faulttolerance.api.CircuitBreakerName;
+import io.smallrye.faulttolerance.autoconfig.FaultToleranceMethod;
+import io.smallrye.faulttolerance.autoconfig.MethodDescriptor;
+
+final class FaultToleranceScanner {
+    private final IndexView index;
+    private final AnnotationStore annotationStore;
+
+    private final AnnotationProxyBuildItem proxy;
+    private final ClassOutput output;
+
+    FaultToleranceScanner(IndexView index, AnnotationStore annotationStore, AnnotationProxyBuildItem proxy,
+            ClassOutput output) {
+        this.index = index;
+        this.annotationStore = annotationStore;
+        this.proxy = proxy;
+        this.output = output;
+    }
+
+    boolean hasFTAnnotations(ClassInfo clazz) {
+        // first check annotations on type
+        if (annotationStore.hasAnyAnnotation(clazz, DotNames.FT_ANNOTATIONS)) {
+            return true;
+        }
+
+        // then check on the methods
+        for (MethodInfo method : clazz.methods()) {
+            if (annotationStore.hasAnyAnnotation(method, DotNames.FT_ANNOTATIONS)) {
+                return true;
+            }
+        }
+
+        // then check on the parent
+        DotName parentClassName = clazz.superName();
+        if (parentClassName == null || parentClassName.equals(DotNames.OBJECT)) {
+            return false;
+        }
+        ClassInfo parentClass = index.getClassByName(parentClassName);
+        if (parentClass == null) {
+            return false;
+        }
+        return hasFTAnnotations(parentClass);
+    }
+
+    void forEachMethod(ClassInfo clazz, Consumer<MethodInfo> action) {
+        for (MethodInfo method : clazz.methods()) {
+            // TODO probably restrict more?
```

Comment:
@mkouba is there perhaps a way to accurately filter out all business methods? Alternatively, I have access to the `BeanInfo`, is there a way to accurately filter out methods that are intercepted (by the Fault Tolerance interceptor ideally)?

Reply:
Unfortunately, it's quite tricky (bridge methods, final methods, methods without bindings, etc.). In theory, we could make the `BeanInfo.getInterceptedMethods()` and relevant types public if it helps...

Reply:
Fair enough. What I have here is OK, it just requires processing more methods than necessary, but doesn't affect correctness.